File: conductor-android-prototype/app/src/main/java/app/conductor/policy/PolicyEngine.kt
package app.conductor.policy

import app.conductor.runtime.AutonomyMode
import app.conductor.runtime.PlanStep
import app.conductor.runtime.PolicyDecision
import app.conductor.runtime.PolicyResult
import app.conductor.runtime.Risk
import app.conductor.runtime.UserPolicy

class PolicyEngine {
    private val sensitiveActions = setOf(
        "outbound_message.send",
        "email.send",
        "public_post.create",
        "calendar.confirm_booking",
        "location.share",
        "contacts.call",
        "purchase.create",
        "banking.transfer.create",
        "payment.send"
    )

    /** Never automate credentials, security settings, or irreversible account destruction. */
    private val blockedAlways = setOf(
        "account_security.change",
        "data.delete",
        "password.enter",
        "mfa.bypass"
    )

    fun evaluate(step: PlanStep, policy: UserPolicy): PolicyResult {
        if (policy.mode == AutonomyMode.ASK_ONLY && step.actionType != "answer.generate") {
            return PolicyResult(
                decision = PolicyDecision.BLOCK,
                reason = "Ask Only mode blocks external actions."
            )
        }

        if (blockedAlways.contains(step.actionType)) {
            return PolicyResult(
                decision = PolicyDecision.BLOCK,
                reason = "This action is never automated by Conductor (security boundary)."
            )
        }

        // Money movement always requires exact approval regardless of autonomy mode.
        if (step.actionType in setOf("purchase.create", "banking.transfer.create", "payment.send")) {
            return PolicyResult(
                decision = PolicyDecision.REQUIRE_APPROVAL,
                reason = "Money-moving actions always require exact user approval of amount and destination."
            )
        }

        if (sensitiveActions.contains(step.actionType)) {
            return PolicyResult(
                decision = PolicyDecision.REQUIRE_APPROVAL,
                reason = "Sensitive external actions require exact user approval."
            )
        }

        if (policy.mode == AutonomyMode.DRAFT_ONLY && step.actionType.endsWith(".create_draft")) {
            return PolicyResult(
                decision = PolicyDecision.ALLOW,
                reason = "Draft creation has no external side effect."
            )
        }

        if (policy.mode == AutonomyMode.DRAFT_ONLY && step.externalSideEffect) {
            return PolicyResult(
                decision = PolicyDecision.REQUIRE_APPROVAL,
                reason = "Draft Only mode requires approval before external side effects."
            )
        }

        if (step.risk == Risk.LOW) {
            return PolicyResult(
                decision = PolicyDecision.ALLOW,
                reason = "Low-risk context or reversible action."
            )
        }

        return PolicyResult(
            decision = PolicyDecision.REQUIRE_APPROVAL,
            reason = "Medium and high-risk steps require user approval."
        )
    }
}

File: conductor-runtime-core/src/policy-engine.js
import { AutonomyMode, Decision, Risk } from "./domain.js";

const SENSITIVE_ACTIONS = new Set([
  "outbound_message.send",
  "email.send",
  "public_post.create",
  "calendar.confirm_booking",
  "location.share"
]);

const BLOCKED_IN_MVP = new Set([
  "purchase.create",
  "account_security.change",
  "data.delete"
]);

export function evaluatePolicy(step, userPolicy) {
  const mode = userPolicy?.mode ?? AutonomyMode.DRAFT_ONLY;

  if (mode === AutonomyMode.ASK_ONLY && step.actionType !== "answer.generate") {
    return {
      decision: Decision.BLOCK,
      reason: "Ask Only mode blocks external actions."
    };
  }

  if (BLOCKED_IN_MVP.has(step.actionType)) {
    return {
      decision: Decision.BLOCK,
      reason: "This action is blocked in the MVP until reliability and trust controls are proven."
    };
  }

  if (SENSITIVE_ACTIONS.has(step.actionType)) {
    return {
      decision: Decision.REQUIRE_APPROVAL,
      reason: "Sensitive external actions require exact user approval."
    };
  }

  if (mode === AutonomyMode.DRAFT_ONLY && step.actionType.endsWith(".create_draft")) {
    return {
      decision: Decision.ALLOW,
      reason: "Draft creation has no external side effect."
    };
  }

  if (mode === AutonomyMode.DRAFT_ONLY && step.externalSideEffect === true) {
    return {
      decision: Decision.REQUIRE_APPROVAL,
      reason: "Draft Only mode requires approval before external side effects."
    };
  }

  if (step.risk === Risk.LOW) {
    return {
      decision: Decision.ALLOW,
      reason: "Low-risk context or reversible action."
    };
  }

  return {
    decision: Decision.REQUIRE_APPROVAL,
    reason: "Medium and high-risk steps require user approval."
  };
}
